### PR TITLE
IBP-4263 fix code inheritance

### PR DESCRIPTION
**Pull request by mderamos**

### CODE REVIEW

File: src/main/java/org/generationcp/middleware/service/impl/GermplasmGroupingServiceImpl.java [hunk 1]
```diff
@@ -117,8 +118,26 @@ public GermplasmGroup getGroupMembers(final Germplasm founder) {
...
+		final List<Germplasm> descendantGroupMembers = new ArrayList<>();
+		this.populateDescendantGroupMembers(descendantGroupMembers, gid, mgid);
+		return descendantGroupMembers;
+	}
+
+	private void populateDescendantGroupMembers(final List<Germplasm> descendantGroupMembers, final Integer gid, final Integer mgid) {
+		// Get advanced children
+		final List<Germplasm> maintenanceAndDerivativeChildren = new ArrayList<>();
+		maintenanceAndDerivativeChildren.addAll(this.germplasmDAO.getChildren(gid, 'M'));
+		maintenanceAndDerivativeChildren.addAll(this.germplasmDAO.getChildren(gid, 'D'));
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/service/impl/GermplasmGroupingServiceImpl.java:131]: Refactor `germplasmDao.getChildren` to `getNonGenerativeChildren(final Integer gid, final List<MethodType> filterMethodTypes)`

File: src/main/java/org/generationcp/middleware/service/impl/GermplasmGroupingServiceImpl.java [hunk 2]
```diff
@@ -117,8 +118,26 @@ public GermplasmGroup getGroupMembers(final Germplasm founder) {
...
-	public List<Germplasm> getGroupMembers(final Integer gid) {
-		return this.germplasmDAO.getManagementGroupMembers(gid);
+	public List<Germplasm> getDescendantGroupMembers(final Integer gid, final Integer mgid) {
+		final List<Germplasm> descendantGroupMembers = new ArrayList<>();
+		this.populateDescendantGroupMembers(descendantGroupMembers, gid, mgid);
+		return descendantGroupMembers;
+	}
+
+	private void populateDescendantGroupMembers(final List<Germplasm> descendantGroupMembers, final Integer gid, final Integer mgid) {
+		// Get advanced children
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/service/impl/GermplasmGroupingServiceImpl.java:128]: Why are we getting only non-generative children? It is not clear from the service name that generative children will be excluded. Maybe we can include in method signature: 
List<Germplasm> getDescendantGroupMembers(Integer gid, Integer mgid, final List<MethodType> filterMethodTypes)

Comment (mderamos) [reply src/main/java/org/generationcp/middleware/service/impl/GermplasmGroupingServiceImpl.java:128]: Hi @darla-leafnode, code inheritance are only applied to advanced children.

File: src/main/java/org/generationcp/middleware/manager/PedigreeDataManagerImpl.java [hunk 3]
```diff
@@ -447,7 +447,7 @@ private GermplasmPedigreeTreeNode getDerivedLines(final GermplasmPedigreeTreeNod
 	}
 
 	private List<Germplasm> getChildren(final Integer gid, final char methodType) {
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/manager/PedigreeDataManagerImpl.java:449]: Should this also be renamed to `getNonGenerativeChildren`?

File: src/main/java/org/generationcp/middleware/dao/GermplasmDAO.java [hunk 4]
```diff
@@ -669,7 +669,7 @@ public long countGroupRelatives(final Integer gid) {
 		return toreturn;
 	}
 
-	public List<Germplasm> getChildren(final Integer gid, final char methodType) {
+	public List<Germplasm> getNonGenerativeChildren(final Integer gid, final char methodType) {
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/dao/GermplasmDAO.java:672]: Sorry, let's rename to `getNonGenerativeDescendants`. Also we don't need a `methodType` parameter here. You can remove Germplasm.GET_DERIVATIVE_CHILDREN and Germplasm.GET_MAINTENANCE_CHILDREN and consolidate the query as:
```
"SELECT {g.*}, {n.*} " + "FROM germplsm g LEFT JOIN names n ON g.gid = n.gid AND n.nstat = 1 "
			+ "JOIN methods m ON g.methn = m.mid AND m.mtype IN ('MAN', 'DER') "
			+ "WHERE g.gnpgs = -1 AND g.gpid2 = :gid and  g.deleted = 0  and g.grplce = 0";
```

File: src/test/java/org/generationcp/middleware/service/impl/GermplasmGroupingServiceImplTest.java [hunk 5]
```diff
@@ -829,23 +826,23 @@ public void testCopyCodedNamesPriorityForSettingPreferredName() {
...
-		final List<Germplasm> groupMembers = new ArrayList<>();
-		Mockito.when(this.germplasmDAO.getManagementGroupMembers(gid)).thenReturn(groupMembers);
-		Assert.assertSame(groupMembers, germplasmGroupingService.getGroupMembers(gid));
+	@Test
+	public void testGetDescendantGroupMembers() {
+		final Integer mgid = 11;
+		final Integer rootGid = 13;
+		final Germplasm derivative = new Germplasm();
+		derivative.setGid(22);
+		derivative.setMgid(11);
```
Comment (darla-leafnode) [inline src/test/java/org/generationcp/middleware/service/impl/GermplasmGroupingServiceImplTest.java:835]: use `mgid` variable

File: src/main/java/org/generationcp/middleware/dao/GermplasmDAO.java (hunk 4 above)
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/dao/GermplasmDAO.java:672]: Sorry, upon going back, I think this should be `getDescendants`

**Review by darla-leafnode — CHANGES_REQUESTED** [review]
Just have some minor comments, please check @mderamos

**Review by darla-leafnode — APPROVED** [review]

**Pull request merged**